python/lsst/ip/isr/calibType.py

```diff
@@ -0,0 +1,630 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+import abc
+import copy
+import datetime
+import yaml
+
+import lsst.afw.table as afwTable
+from lsst.log import Log
+from lsst.daf.base import PropertyList
+from astropy.table import Table
+
+__all__ = ["IsrCalib", "IsrProvenance"]
+
+
+class IsrCalib(abc.ABC):
+    """Generic calibration type.
+
+    Subclasses must implement the toDict, fromDict, toTable, fromTable
+    methods that allow the calibration information to be converted
+    from dictionaries and afw tables.  This will allow the calibration
+    to be persisted using the base class read/write methods.
+
+    The validate method is intended to provide a common way to check
+    that the calibration is valid (internally consistent) and
+    appropriate (usable with the intended data).  The apply method is
+    intended to allow the calibration to be applied in a consistent
+    manner.
+
+    Parameters
+    ----------
+    detectorName : `str`, optional
+        Name of the detector this calibration is for.
+    detectorSerial : `str`, optional
+        Identifier for the detector.
+    detector : `lsst.afw.cameraGeom.Detector`, optional
+        Detector to extract metadata from.
+    log : `lsst.log.Log`, optional
+        Log for messages.
+
+    """
+    _OBSTYPE = 'generic'
+    _SCHEMA = 'NO SCHEMA'
+    _VERSION = 0
+
+    def __init__(self, detectorName=None, detectorSerial=None, detector=None, log=None, **kwargs):
+        self._detectorName = detectorName
+        self._detectorSerial = detectorSerial
+        self._metadata = PropertyList()
+
+        # Ensure that we have the obs type required by calibration ingest
+        self._metadata["OBSTYPE"] = self._OBSTYPE
+        self._metadata[self._OBSTYPE + "_SCHEMA"] = self._SCHEMA
+        self._metadata[self._OBSTYPE + "_VERSION"] = self._VERSION
+
+        self.log = log if log else Log.getLogger("lsst.ip.isr.IsrCalib")
+
+        if detector:
+            self.fromDetector(detector)
+        self.updateMetadata(setDate=False)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}(obstype={self._OBSTYPE}, detector={self._detectorName}, )"
+
+    def __eq__(self, other):
+        """Calibration equivalence.
+
+        Subclasses will need to check specific sub-properties.  The
+        default is only to check common entries.
+        """
+        if not isinstance(other, self.__class__):
+            return False
+
+        for attr in ('_OBSTYPE', '_SCHEMA', '_VERSION',
+                     '_detectorName', '_detectorSerial', '_metadata'):
```

Comment:
It might help to make this second line of attributes be a class property so that subclasses can set the one list and then reuse this `__eq__` implementation.